TailLight/device.cpp

```diff
@@ -1,50 +1,186 @@
     if (!NT_SUCCESS(status)) {
-        KdPrint(("WdfTimerCreate failed 0x%x\n", status));
+        KdPrint(("TailLight: WdfDeviceAllocAndQueryProperty"
+            "DevicePropertyPhysicalDeviceObjectName failed 0x%x\n", status));
         return status;
     }
 
-    status = WdfTimerStart(timer, 0); // no wait
+    // initialize pDeviceContext->PdoName based on memory
+    size_t bufferLength = 0;
+    pdoName.Buffer = (WCHAR*)WdfMemoryGetBuffer(memory, &bufferLength);
+    if (pdoName.Buffer == NULL)
+        return STATUS_UNSUCCESSFUL;
+
+    pdoName.MaximumLength = (USHORT)bufferLength;
+    pdoName.Length = (USHORT)bufferLength - sizeof(UNICODE_NULL);
+
+    KdPrint(("TailLight: PdoName: %wZ\n", pdoName)); // outputs "\Device\00000083
+
+    return status;
+}
+
+NTSTATUS AllocateHwIdString(WDFDEVICE device, UNICODE_STRING& hwId)
+{
+    WDF_OBJECT_ATTRIBUTES attributes = {};
+    WDF_OBJECT_ATTRIBUTES_INIT(&attributes);
+    attributes.ParentObject = device; // auto-delete with device
+    NTSTATUS status = STATUS_SUCCESS;
+
+    WDFMEMORY memory = 0;
+    status = WdfDeviceAllocAndQueryProperty(device,
+        DevicePropertyHardwareID,
+        NonPagedPoolNx,
+        &attributes,
+        &memory);
     if (!NT_SUCCESS(status)) {
-        KdPrint(("WdfTimerStart failed 0x%x\n", status));
+        KdPrint(("TailLight: WdfDeviceAllocAndQueryProperty"
+            "DevicePropertyHardwareID failed 0x%x\n", 
+            status));
         return status;
     }
 
+    hwId.Buffer = (PWCHAR)WdfMemoryGetBuffer(memory, (SIZE_T*) & (hwId.MaximumLength));
+    if (hwId.Buffer == nullptr)
+        return STATUS_UNSUCCESSFUL;
+
+    hwId.Length = (USHORT)hwId.MaximumLength - sizeof(UNICODE_NULL);
+
+    KdPrint(("TailLight: Device hardware ID %S\n", hwId.Buffer));
+
     return status;
 }
 
+// TODO: Consider passing in the property and making a generic property retrieval
+NTSTATUS AllocateHwIdString(WDFIOTARGET target, UNICODE_STRING& hwId)
+/*++
+Routine Description:
+    Creates from non-paged memory a MULTI_SZ unicode hardware Id of the I/O 
+    target and sets the default object lifetime to the IO target.
+
+Arguments:
+    WDFIOTARGET - A WDF IO target
+
+    STRING - Where to place the allocated string.
+--*/
+{
+    WDF_OBJECT_ATTRIBUTES attributes = {};
+    WDF_OBJECT_ATTRIBUTES_INIT(&attributes);
+    attributes.ParentObject = target; // auto-delete with device
+    NTSTATUS status = STATUS_SUCCESS;
+
+    WDFMEMORY memory = 0;
+    status = WdfIoTargetAllocAndQueryTargetProperty(target,
+        DevicePropertyHardwareID,
+        NonPagedPoolNx,
+        &attributes,
+        &memory);
+    if (!NT_SUCCESS(status)) {
+        KdPrint(("TailLight: WdfDeviceAllocAndQueryProperty"
+            "DevicePropertyHardwareID failed 0x%x\n",
+            status));
+        return status;
+    }
+
+    hwId.Buffer = (PWCHAR)WdfMemoryGetBuffer(memory, (SIZE_T*) & (hwId.MaximumLength));
```

Comment:
Please use a `size_t bufferLength` output variable here for configuring the `Length` and `MaximumLength` members like already done for `DevicePropertyPhysicalDeviceObjectName`.